addons/mail/static/src/messaging/entity/thread/thread.js

```diff
@@ -0,0 +1,674 @@
+            if (!cache) {
+                cache = this.env.entities.ThreadCache.create({
+                    stringifiedDomain,
+                    thread: this,
+                });
+            }
+            return cache;
+        }
+
+        /**
+         * @returns {mail.messaging.entity.ChatWindow[]}
+         */
+        get chatWindows() {
+            const chatWindowViewers = this.viewers.filter(viewer => !!viewer.chatWindow);
+            return chatWindowViewers.map(viewer => viewer.chatWindow);
+        }
+
+        /**
+         * @returns {string}
+         */
+        get displayName() {
+            if (this.channel_type === 'chat' && this.directPartner) {
+                return this.custom_channel_name || this.directPartner.nameOrDisplayName;
+            }
+            if (this.channel_type === 'livechat') {
+                // FIXME: should be patch in im_livechat
+                return this.correspondent_name;
+            }
+            return this.name;
+        }
+
+        /**
+         * Fetch attachments linked to a record. Useful for populating the store
+         * with these attachments, which are used by attachment box in the chatter.
+         */
+        async fetchAttachments() {
+            const attachmentsData = await this.env.rpc({
+                model: 'ir.attachment',
+                method: 'search_read',
+                domain: [
+                    ['res_id', '=', this.id],
+                    ['res_model', '=', this.model],
+                ],
+                fields: ['id', 'name', 'mimetype'],
+                orderBy: [{ name: 'id', asc: false }],
+            });
+            for (const attachmentData of attachmentsData) {
+                this.env.entities.Attachment.insert(Object.assign({
+                    res_id: this.id,
+                    res_model: this.model,
+                }, attachmentData));
+            }
+            this.update({ areAttachmentsLoaded: true });
+        }
+
+        /**
+         * @returns {boolean}
+         */
+        get isModeratedByUser() {
+            if (this.model !== 'mail.channel') {
+                return false;
+            }
+            return Thread.moderatedChannelIds.includes(this.id);
+        }
+
+        /**
+         * @returns {mail.messaging.entity.Message|undefined}
+         */
+        get lastMessage() {
+            return this.mainCache.lastMessage;
+        }
+
+        loadNewMessages() {
+            this.mainCache.loadNewMessages();
+        }
+
+        /**
+         * @returns {mail.messaging.entity.ThreadCache}
+         */
+        get mainCache() {
+            return this.caches.find(cache => cache.stringifiedDomain === '[]');
+        }
+
+        /**
+         * Mark the specified conversation as read/seen.
+         */
+        async markAsSeen() {
+            if (this.message_unread_counter === 0) {
+                return;
+            }
```

Comment:
Do we really need this preemptive return? Seems like we shouldn't reach this point if that is not the case.